daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/LimitsSpec.scala

```diff
@@ -0,0 +1,436 @@
+              templateId shouldBe T
+              choiceName shouldBe "NoOp"
+              limit shouldBe limit
+              false
+          }
+        }
+      } shouldBe 2
+    }
+
+    "refuse to build a transaction with too many input contracts" in {
+      val limit = 10
+      val limits = interpretation.Limits.Lenient.copy(transactionInputContracts = limit)
+      val signatories = committers.take(1)
+      val contract = mkContract(signatories, Set.empty)
+      val contracts = (_: Value.ContractId) => contract
+      val cids = (1 to 99).map(i => Value.ContractId.V1(crypto.Hash.hashPrivateKey(s"contract$i")))
+      val e = e"Mod:fetches"
+      List(1, limit, limit + 1, 49).count { i =>
+        val result = eval(limits, contracts, committers, e, asSCids(cids.take(i)))
+        if (i <= limit) {
+          result shouldBe a[Right[_, _]]
+          true
+        } else {
+          inside(result) {
+            case Left(
+                  SError.SErrorDamlException(IE.Limit(IE.Limit.TransactionInputContracts(limit)))
+                ) =>
+              limit shouldBe limit
+              false
+          }
+        }
+      } shouldBe 2
+
+    }
+
+  }
+
+  private def asSParties(parties: Iterable[Ref.Party]) =
+    SValue.SList(parties.map(SValue.SParty).to(FrontStack))
+
+  private def asSCids(cids: Iterable[Value.ContractId]) =
+    SValue.SList(cids.map(SValue.SContractId).to(FrontStack))
+
+  private val txSeed = crypto.Hash.hashPrivateKey(this.getClass.getCanonicalName)
+
+  private def typeAndCompile(pkg: Ast.Package): PureCompiledPackages = {
+    import defaultParserParameters.defaultPackageId
+    val rawPkgs = Map(defaultPackageId -> pkg)
+    Validation.checkPackage(PackageInterface(rawPkgs), defaultPackageId, pkg)
+    val compilerConfig = Compiler.Config.Dev
+    PureCompiledPackages.assertBuild(rawPkgs, compilerConfig)
+  }
+
+  private[this] val T = { val Ast.TTyCon(t) = t"Mod:T"; t }
+
+  private[this] val aCid = Value.ContractId.V1(crypto.Hash.hashPrivateKey("a contract ID"))
+  private[this] def mkContract(signatories: Iterable[Ref.Party], observers: Iterable[Ref.Party]) = {
+
+    Versioned(
+      TransactionVersion.StableVersions.max,
+      Value.ContractInstance(
+        T,
+        Value.ValueRecord(
+          None,
+          ImmArray(
+            None -> Value.ValueList(signatories.view.map(Value.ValueParty).to(FrontStack)),
+            None -> Value.ValueList(observers.view.map(Value.ValueParty).to(FrontStack)),
+          ),
+        ),
+        "Agreement",
+      ),
+    )
+  }
+
+  private[this] def eval(
+      limits: interpretation.Limits,
+      contracts: Value.ContractId => Versioned[Value.ContractInstance],
+      committers: Set[Ref.Party],
+      e: Ast.Expr,
+      agrs: SValue*
+  ): Either[SError.SError, SubmittedTransaction] = {
+    val updateSE =
+      SExpr.SEApp(pkgs.compiler.unsafeCompile(e), agrs.view.map(SExpr.SEValue(_)).toArray)
+    val machine = Speedy.Machine(
```

Comment:
this looks like `fromUpdateSExpr` or am I missing something?

Reply:
Actually it was not, `fromUpdateSExpr` took only a single committers. I change the method to take set of parties instead without too much issues, it was used less than what I expected.